<a href="https://colab.research.google.com/github/idouglas237020/CS4410/blob/main/Ingram_Douglas_Homework_7.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import tensorflow as tf
from tensorflow import keras
import matplotlib.pyplot as plt
import time

# Exercise 16.1 - Fashion-MNIST vs. MNIST

This exercise compares the performance and training time of the convolutional neural network on the Fashion-MNIST and MNIST datasets.

In [19]:
fashion_mnist = keras.datasets.fashion_mnist

(X_train, y_train), (X_test, y_test) = fashion_mnist.load_data()

X_train = X_train.astype("float32") / 255
X_test = X_test.astype("float32") / 255

X_train = X_train.reshape(-1, 28, 28, 1)
X_test = X_test.reshape(-1, 28, 28, 1)

print("Training shape:", X_train.shape)
print("Testing shape:", X_test.shape)

Training shape: (60000, 28, 28, 1)
Testing shape: (10000, 28, 28, 1)


In [10]:
from tensorflow.keras import models, layers

model = models.Sequential([
    layers.Conv2D(64, (3, 3), activation='relu', input_shape=(28, 28, 1)),
    layers.MaxPooling2D((2, 2)),
    layers.Conv2D(128, (3, 3), activation='relu'),
    layers.MaxPooling2D((2, 2)),
    layers.Flatten(),
    layers.Dense(128, activation='relu'),
    layers.Dense(10, activation='softmax')
])

model.summary()

/usr/local/lib/python3.13/dist-packages/keras/src/layers/convolutional/base_conv.py:113: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Model: "sequential_1"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ conv2d_2 (Conv2D)               │ (None, 26, 26, 64)     │           640 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ max_pooling2d_2 (MaxPooling2D)  │ (None, 13, 13, 64)     │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ conv2d_3 (Conv2D)               │ (None, 11, 11, 128)    │        73,856 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ max_pooling2d_3 (MaxPooling2D)  │ (None, 5, 5, 128)      │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ flatten_1 (Flatten)             │ (None, 3200)           │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_2 (Dense)                 │ (None, 128)            │       409,728 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_3 (Dense)                 │ (None, 10)             │         1,290 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 485,514 (1.85 MB)

 Trainable params: 485,514 (1.85 MB)

 Non-trainable params: 0 (0.00 B)

## Fashion-MNIST Training

In [20]:
model.compile(
    optimizer='adam',
    loss='sparse_categorical_crossentropy',
    metrics=['accuracy']
)

start_time = time.time()

history = model.fit(
    X_train,
    y_train,
    epochs=5,
    batch_size=64,
    validation_split=0.1
)

training_time = time.time() - start_time

print("Training time:", training_time, "seconds")

Epoch 1/5
844/844 ━━━━━━━━━━━━━━━━━━━━ 105s 123ms/step - accuracy: 0.8424 - loss: 0.4486 - val_accuracy: 0.8868 - val_loss: 0.2977
Epoch 2/5
844/844 ━━━━━━━━━━━━━━━━━━━━ 101s 120ms/step - accuracy: 0.9016 - loss: 0.2663 - val_accuracy: 0.9010 - val_loss: 0.2739
Epoch 3/5
844/844 ━━━━━━━━━━━━━━━━━━━━ 143s 121ms/step - accuracy: 0.9205 - loss: 0.2167 - val_accuracy: 0.9102 - val_loss: 0.2545
Epoch 4/5
844/844 ━━━━━━━━━━━━━━━━━━━━ 143s 123ms/step - accuracy: 0.9304 - loss: 0.1854 - val_accuracy: 0.9140 - val_loss: 0.2479
Epoch 5/5
844/844 ━━━━━━━━━━━━━━━━━━━━ 103s 123ms/step - accuracy: 0.9418 - loss: 0.1597 - val_accuracy: 0.9167 - val_loss: 0.2332
Training time: 596.1826691627502 seconds


In [21]:
test_loss, test_accuracy = model.evaluate(X_test, y_test, verbose=0)

print("Fashion-MNIST test accuracy:", test_accuracy)
print("Fashion-MNIST training time:", training_time, "seconds")

Fashion-MNIST test accuracy: 0.9106000065803528
Fashion-MNIST training time: 596.1826691627502 seconds


## MNIST Comparison

In [22]:
from tensorflow.keras.datasets import mnist

(X_train, y_train), (X_test, y_test) = mnist.load_data()

X_train = X_train.astype("float32") / 255
X_test = X_test.astype("float32") / 255

X_train = X_train.reshape(-1, 28, 28, 1)
X_test = X_test.reshape(-1, 28, 28, 1)

print("Training shape:", X_train.shape)
print("Testing shape:", X_test.shape)

Training shape: (60000, 28, 28, 1)
Testing shape: (10000, 28, 28, 1)


In [23]:
model_mnist = models.Sequential([
    layers.Conv2D(64, (3, 3), activation='relu', input_shape=(28, 28, 1)),
    layers.MaxPooling2D((2, 2)),
    layers.Conv2D(128, (3, 3), activation='relu'),
    layers.MaxPooling2D((2, 2)),
    layers.Flatten(),
    layers.Dense(128, activation='relu'),
    layers.Dense(10, activation='softmax')
])

model_mnist.compile(
    optimizer='adam',
    loss='sparse_categorical_crossentropy',
    metrics=['accuracy']
)

In [24]:
model.compile(
    optimizer='adam',
    loss='sparse_categorical_crossentropy',
    metrics=['accuracy']
)

start_time = time.time()

history_mnist = model_mnist.fit(
    X_train,
    y_train,
    epochs=5,
    batch_size=64,
    validation_split=0.1
)

training_time = time.time() - start_time

print("Training time:", training_time, "seconds")

Epoch 1/5
844/844 ━━━━━━━━━━━━━━━━━━━━ 104s 122ms/step - accuracy: 0.9580 - loss: 0.1372 - val_accuracy: 0.9875 - val_loss: 0.0447
Epoch 2/5
844/844 ━━━━━━━━━━━━━━━━━━━━ 102s 121ms/step - accuracy: 0.9871 - loss: 0.0408 - val_accuracy: 0.9892 - val_loss: 0.0366
Epoch 3/5
844/844 ━━━━━━━━━━━━━━━━━━━━ 142s 122ms/step - accuracy: 0.9910 - loss: 0.0282 - val_accuracy: 0.9907 - val_loss: 0.0327
Epoch 4/5
844/844 ━━━━━━━━━━━━━━━━━━━━ 102s 121ms/step - accuracy: 0.9938 - loss: 0.0193 - val_accuracy: 0.9907 - val_loss: 0.0369
Epoch 5/5
844/844 ━━━━━━━━━━━━━━━━━━━━ 142s 122ms/step - accuracy: 0.9952 - loss: 0.0145 - val_accuracy: 0.9915 - val_loss: 0.0345
Training time: 632.9511859416962 seconds


In [25]:
test_loss, test_accuracy = model_mnist.evaluate(X_test, y_test, verbose=0)

print("MNIST test accuracy:", test_accuracy)
print("MNIST training time:", training_time, "seconds")

MNIST test accuracy: 0.9908999800682068
MNIST training time: 632.9511859416962 seconds


The convolutional neural network achieved about 91.06% test accuracy on the Fashion-MNIST dataset and about 99.09% test accuracy on the MNIST dataset. Fashion-MNIST took approximately 596.18 seconds to train, while MNIST took approximately 632.95 seconds. The training times were fairly similar because both datasets contain 28 × 28 grayscale images and used the same model settings. The model performed better on MNIST because handwritten digits are generally easier to distinguish than the clothing categories in Fashion-MNIST.

# Exercise 16.4 - Convnet Layers

This exercise compares the original convolutional neural network with two modified versions: one with the first Dense layer removed and one with an added Dense layer containing 4096 neurons. The models are compared by test accuracy and training time.

In [13]:
from tensorflow.keras import models, layers

model_no_dense = models.Sequential([
    layers.Conv2D(64, (3, 3), activation='relu', input_shape=(28, 28, 1)),
    layers.MaxPooling2D((2, 2)),
    layers.Conv2D(128, (3, 3), activation='relu'),
    layers.MaxPooling2D((2, 2)),
    layers.Flatten(),
    layers.Dense(10, activation='softmax')
])

model_no_dense.compile(
    optimizer='adam',
    loss='sparse_categorical_crossentropy',
    metrics=['accuracy']
)

In [14]:
start_time = time.time()

history_no_dense = model_no_dense.fit(
    X_train,
    y_train,
    epochs=5,
    batch_size=64,
    validation_split=0.1
)

no_dense_time = time.time() - start_time

print("Training time without first Dense layer:", no_dense_time, "seconds")

Epoch 1/5
844/844 ━━━━━━━━━━━━━━━━━━━━ 103s 121ms/step - accuracy: 0.9491 - loss: 0.1725 - val_accuracy: 0.9802 - val_loss: 0.0712
Epoch 2/5
844/844 ━━━━━━━━━━━━━━━━━━━━ 98s 117ms/step - accuracy: 0.9831 - loss: 0.0536 - val_accuracy: 0.9870 - val_loss: 0.0449
Epoch 3/5
844/844 ━━━━━━━━━━━━━━━━━━━━ 143s 118ms/step - accuracy: 0.9877 - loss: 0.0387 - val_accuracy: 0.9880 - val_loss: 0.0416
Epoch 4/5
844/844 ━━━━━━━━━━━━━━━━━━━━ 142s 118ms/step - accuracy: 0.9906 - loss: 0.0299 - val_accuracy: 0.9903 - val_loss: 0.0356
Epoch 5/5
844/844 ━━━━━━━━━━━━━━━━━━━━ 142s 117ms/step - accuracy: 0.9929 - loss: 0.0230 - val_accuracy: 0.9903 - val_loss: 0.0328
Training time without first Dense layer: 627.8449599742889 seconds


In [15]:
no_dense_loss, no_dense_accuracy = model_no_dense.evaluate(
    X_test, y_test, verbose=0
)

print("Test accuracy without first Dense layer:", no_dense_accuracy)
print("Training time without first Dense layer:", no_dense_time, "seconds")

Test accuracy without first Dense layer: 0.9891999959945679
Training time without first Dense layer: 627.8449599742889 seconds


In [16]:
model_4096 = models.Sequential([
    layers.Conv2D(64, (3, 3), activation='relu', input_shape=(28, 28, 1)),
    layers.MaxPooling2D((2, 2)),
    layers.Conv2D(128, (3, 3), activation='relu'),
    layers.MaxPooling2D((2, 2)),
    layers.Flatten(),
    layers.Dense(4096, activation='relu'),
    layers.Dense(128, activation='relu'),
    layers.Dense(10, activation='softmax')
])

model_4096.compile(
    optimizer='adam',
    loss='sparse_categorical_crossentropy',
    metrics=['accuracy']
)

In [17]:
start_time = time.time()

history_4096 = model_4096.fit(
    X_train,
    y_train,
    epochs=5,
    batch_size=64,
    validation_split=0.1
)

time_4096 = time.time() - start_time

print("Training time with 4096-neuron Dense layer:", time_4096, "seconds")

Epoch 1/5
844/844 ━━━━━━━━━━━━━━━━━━━━ 350s 413ms/step - accuracy: 0.9625 - loss: 0.1202 - val_accuracy: 0.9865 - val_loss: 0.0436
Epoch 2/5
844/844 ━━━━━━━━━━━━━━━━━━━━ 344s 408ms/step - accuracy: 0.9878 - loss: 0.0386 - val_accuracy: 0.9855 - val_loss: 0.0505
Epoch 3/5
844/844 ━━━━━━━━━━━━━━━━━━━━ 350s 414ms/step - accuracy: 0.9922 - loss: 0.0256 - val_accuracy: 0.9905 - val_loss: 0.0361
Epoch 4/5
844/844 ━━━━━━━━━━━━━━━━━━━━ 354s 419ms/step - accuracy: 0.9942 - loss: 0.0185 - val_accuracy: 0.9895 - val_loss: 0.0384
Epoch 5/5
844/844 ━━━━━━━━━━━━━━━━━━━━ 376s 413ms/step - accuracy: 0.9946 - loss: 0.0165 - val_accuracy: 0.9913 - val_loss: 0.0329
Training time with 4096-neuron Dense layer: 1774.2377269268036 seconds


In [18]:
loss_4096, accuracy_4096 = model_4096.evaluate(
    X_test, y_test, verbose=0
)

print("Test accuracy with 4096-neuron Dense layer:", accuracy_4096)
print("Training time with 4096-neuron Dense layer:", time_4096, "seconds")

Test accuracy with 4096-neuron Dense layer: 0.9916999936103821
Training time with 4096-neuron Dense layer: 1774.2377269268036 seconds


Exercise 16.4 Results:
Removing the first Dense layer reduced the test accuracy from about 99.09% to 98.92%. The training time also changed from about 632.95 seconds to 627.84 seconds. Adding a Dense layer with 4096 neurons increased the test accuracy to about 99.17%, but it also increased the training time significantly to about 1774.24 seconds. The larger Dense layer produced only a small improvement in accuracy while requiring much more training time, so the original model provided a better balance between accuracy and speed.
